pkg/tracing/tracing.go

```diff
@@ -48,20 +176,97 @@ type (
 
 	// Tracer is the tracer.
 	Tracer struct {
-		tracer *zipkingo.Tracer
-		tags   map[string]string
-		closer io.Closer
+		trace.Tracer
+		tp         *sdktrace.TracerProvider
+		propagator propagation.TextMapPropagator
 	}
 
-	noopCloser struct{}
+	// Span is the span of the Tracing.
+	Span struct {
+		trace.Span
+		tracer *Tracer
+		ctx    context.Context
+	}
+)
+
+const (
+	//see: https://www.w3.org/TR/trace-context/
+	headerFormatTraceContext = "trace-context"
+	headerFormatB3           = "b3"
+
+	jaegerModeAgent     jaegerMode = "agent"
+	jaegerModeCollector jaegerMode = "collector"
+
+	otlpProtocolHTTP otlpProtocol = "http"
+	otlpProtocolGRPC otlpProtocol = "grpc"
 )
 
+func (spec *Spec) UnmarshalJSON(data []byte) error {
+	type innerSpec Spec
+	inner := &innerSpec{
+		SpanLimits: &SpanLimitsSpec{
+			AttributeValueLengthLimit:   sdktrace.DefaultAttributeValueLengthLimit,
+			AttributeCountLimit:         sdktrace.DefaultAttributeCountLimit,
+			EventCountLimit:             sdktrace.DefaultEventCountLimit,
+			LinkCountLimit:              sdktrace.DefaultLinkCountLimit,
+			AttributePerEventCountLimit: sdktrace.DefaultAttributePerEventCountLimit,
+			AttributePerLinkCountLimit:  sdktrace.DefaultAttributePerLinkCountLimit,
+		},
+		SampleRate: 1,
+		BatchLimits: &BatchLimitsSpec{
+			MaxQueueSize:       sdktrace.DefaultMaxQueueSize,
+			BatchTimeout:       sdktrace.DefaultScheduleDelay,
+			ExportTimeout:      sdktrace.DefaultExportTimeout,
+			MaxExportBatchSize: sdktrace.DefaultMaxExportBatchSize,
+		},
+		HeaderFormat: headerFormatTraceContext,
+	}
+
+	if err := json.Unmarshal(data, inner); err != nil {
+		return err
+	}
+
+	*spec = Spec(*inner)
+	return nil
+
+}
+
+func (spec *Spec) Validate() error {
+	if spec.Exporter == nil && spec.Zipkin == nil {
+		return fmt.Errorf("exporter and zipkin cannot both be empty")
+	}
+
+	if spec.Exporter != nil && spec.Zipkin != nil {
+		return fmt.Errorf("exporter and zipkin cannot exist at the same time")
+	}
+
+	return nil
+}
+
 // Validate validates Spec.
-func (spec *ZipkinSpec) Validate() error {
-	if spec.Hostport != "" {
-		_, err := zipkingo.NewEndpoint("", spec.Hostport)
-		if err != nil {
-			return err
+func (spec *ExporterSpec) Validate() error {
+
+	if spec == nil {
+		return fmt.Errorf("exporter cannot be empty")
+	}
+
+	if spec.Jaeger == nil && spec.Zipkin == nil && spec.OTLP == nil {
+		return fmt.Errorf("exporter cannot be empty")
+	}
+
+	return nil
+}
+
+// Validate validates Spec.
+func (spec *JaegerSpec) Validate() error {
+
```

Comment:
per coding style, please remove these blank lines, and the blank lines before the closing `}` of a function.